flink-table-store-core/src/main/java/org/apache/flink/table/store/file/predicate/PredicateConverter.java

```diff
@@ -131,22 +139,28 @@ private Predicate visitBiFunction(
         return Optional.empty();
     }
 
-    private Optional<Literal> extractLiteral(Expression expression) {
+    private Optional<Literal> extractLiteral(DataType expectedType, Expression expression) {
+        Literal literal = null;
         if (expression instanceof ValueLiteralExpression) {
             ValueLiteralExpression valueExpression = (ValueLiteralExpression) expression;
-            DataType type = valueExpression.getOutputDataType();
-            return supportsPredicate(type.getLogicalType())
-                    ? Optional.of(
-                            new Literal(
-                                    type.getLogicalType(),
-                                    getConverter(type)
-                                            .toInternalOrNull(
-                                                    valueExpression
-                                                            .getValueAs(type.getConversionClass())
-                                                            .get())))
-                    : Optional.empty();
+            LogicalType expectedLogicalType = expectedType.getLogicalType();
+            DataType actualType = valueExpression.getOutputDataType();
+            LogicalType actualLogicalType = actualType.getLogicalType();
+            DataStructureConverter<Object, Object> converter = getConverter(expectedType);
+            Object value = valueExpression.getValueAs(actualType.getConversionClass()).get();
+            if (!actualLogicalType.equals(expectedLogicalType)
+                    && supportsImplicitCast(actualLogicalType, expectedLogicalType)) {
+                try {
+                    value = TypeUtils.castFromString(value.toString(), expectedLogicalType);
+                } catch (UnsupportedOperationException e) {
```

Comment:
Let catch `Exception ignore`?
I don't think we want to throw an exception here no matter what the situation is.

Reply:
> Let catch Exception ignore?
I don't think we want to throw an exception here no matter what the situation is.

the reason catching exception here is to let `PredicateTest#testUnsupportedExpression` pass. If not caught here, once the casting fails, it'll directly throw `UnsupportedOperationException`, other than `PredicateConverter.UnsupportedExpression`. I think it depends on whether we want an unified exception thrown by `PredicateConverter`

Reply:
finally got ya